### PR TITLE
absorbed mode MLA RFC

### PR DESCRIPTION
**Pull request by Bob-Chen222**
## Motivation
DeepSeek-V2/V3 and subsequent MLA-based models rely on the absorbed-path formulation for tractable inference — the KV cache shrinks from per-head [K,V] to a shared single-head latent [c_kv ‖ k_pe], and the Q/O projections factor through a low-rank bottleneck. sglang-jax currently only has an MHA backend flashattention_backend.py and MHATokenToKVPool memory_pool.py, neither of which fits. This RFC designs the MLA backend + pool, taking tpu-inference and sglang-GPU as reference points.

## Checklist

- [ ] Please use English, otherwise it will be closed.
- [ ] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: docs/design/MLA.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+                                 self.forward_metadata.page_indices,
+                                 self.forward_metadata.seq_lens,
+                                 self.forward_metadata.distribution)
+```
+
+What this costs: two small new files. Every method has one code path. Metadata carries exactly the fields the MLA kernel needs. Sharding specs are fixed at the class level. The MHA path is untouched, so MHA regressions from this change are impossible by construction.
+
+#### 3.2.C Verdict
+
+Proceed with the new-class design (3.2.B). The extend-in-place design (3.2.A) turns the shared classes into multiplexers with near-zero shared code, pollutes every caller with rank- and layout-dependent branches, and keeps inapplicable metadata fields alive. The cost of the split is one file (`mla_backend.py`) and one class (`MLATokenToKVPool`); both are naturally-bounded scopes. Dispatch happens once in `model_runner.py` (§3.10).
```
Comment (zhengkezhou1) [inline docs/design/MLA.md:173]: +1 on Option B. We tried Option A on our experimental branch and confirmed the two paths share almost no code — a separate class is cleaner.

**Review by zhengkezhou1 — COMMENTED** [review]
Hi, we did some research and experimental implementation based on [v1](https://github.com/vllm-project/tpu-inference/pull/1022), [v2](https://github.com/vllm-project/tpu-inference/pull/1971). Our work aligns with this RFC, and we'd like to sync up to avoid duplication.

 ## Current Progress

**Projection Layer (MLAAttention)**
- Q-side absorbs W_k, KV-side skips kv_b_proj
- Kernel output goes through W_v restoration then o_proj

**Kernels**
- Ported `ref_mla_ragged_paged_attention` (pure JAX reference implementation)
- Ported Pallas kernel `mla_ragged_paged_attention` with FlashAttention-2 online softmax
- Includes cache update (`update_kv_cache`) and cache shape computation 

**Tests**
- Layered verification: projection equivalence → projection+kernel → full pipeline
- TPU validated with empirical tolerances (CPU: atol=1e-5, TPU: atol=0.02 due to MXU bf16 precision)
- Ref kernel vs Pallas kernel consistency tests

## On Architecture

Our current adapter layer adds an MLA branch inside FlashAttention
(corresponding to RFC §3.2.A). After reading the RFC's analysis, **we agree
that Option B (a separate MLAAttentionBackend) is the cleaner approach** and
are happy to align with it.

## Division of Work

We've been working on the projection layer and ref kernel. We'd like to
discuss ownership of `MLAAttentionBackend`, `MLATokenToKVPool`, and
ModelRunner dispatch. We should also align on interface contracts first — what
 format the projection layer outputs to the backend, and how the backend
interacts with the pool.

File: docs/design/MLA.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+### 3.5 Kernel interface divergence (RPA v3 vs. MLA v2)
+
+| Field | RPA v3 | MLA v2 |
+|---|---|---|
+| Q inputs | `queries [T, n_h, head_dim]` — single | `ql_nope [T, n_h, lkv_dim] + q_pe [T, n_h, r_dim]` — split, `ql_nope` is absorbed |
+| Current K | `keys [T, n_kv, head_dim]` — per-head | `new_kv_c [T, lkv_dim]` — single-head latent |
+| Current V | `values [T, n_kv, head_dim]` — per-head | `new_k_pe [T, r_dim]` — rotated key, no V |
+| Cache | `[pages, page_size, n_kv*2, head_dim]` — interleaved | `[pages, page_size_per_kv_packing, kv_packing, align_to(lkv+r_dim, 128)]` — concat |
+| Shared metadata | `kv_lens`, `page_indices`, `cu_q_lens`, `distribution` | Same |
+| `cu_kv_lens` | ✓ | ✗ dropped |
```
Comment (JamesBrianD) [inline docs/design/MLA.md]: Without `cu_kv_lens`, how do different sequences index the start and end pages?

File: docs/design/MLA.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+`MLAAttentionMetadata`: `num_seqs`, `cu_q_lens`, `page_indices`, `seq_lens`, `distribution`.
+
+`__call__(payload, layer, forward_batch, pool)`:
+
+1. Unpack `(ql_nope, q_pe, new_kv_c, new_k_pe)` from the layer.
+2. Pad `new_k_pe` and `q_pe` from 64 to 128 to satisfy `r_dim % 128 == 0`.
+3. Write `[new_kv_c ‖ new_k_pe]` into the MLA pool; the kernel handles this via its `updated_cache_kv` return.
+4. Call `mla_ragged_paged_attention(ql_nope, q_pe, new_kv_c, new_k_pe, cache_kv, …)`.
+5. Return `(o_latent, updated_cache_kv)`.
```
Comment (JamesBrianD) [inline docs/design/MLA.md:260]: By analogy with RPA v3, using the MLA kernel may also require utilizing `cu_kv_lens` to determine the start and end indices for the KV pairs.

File: docs/design/MLA.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+| Cache layout | 3D: `(size + page_size, 1, kv_lora_rank + qk_rope_head_dim)` — simple fused | 4D packed: `(pages, page_size/packing, packing, align_to(kv_dim, 128))` | 4D packed (same family as tpu-inference) |
+| Absorb path | **Factored at runtime** — `w_kc`, `w_vc` applied via `bmm` in forward | **Pre-fused post-load** — `W_UK_T`, `W_UV` quantized and absorbed into weights; original `kv_b_proj` deleted | **Factored at runtime** (same as sglang-GPU) — see §3.3 FLOPs argument |
+| Kernel | FlashMLA (custom CUDA w/ tile scheduler), FlashInfer `BatchMLAPagedAttentionWrapper`, cutlass | `mla_ragged_paged_attention` Pallas v2 | Same Pallas v2 kernel as tpu-inference |
+| RoPE on `k_pe` | In forward pass before write | Applied pre-write or in kernel | Applied on write (pre-rotated before cache update) |
+
+Where we diverge from each reference and why:
+
+- **vs. tpu-inference on absorb**: tpu-inference pre-fuses `W_UK_T` and `W_UV` into the weights after loading. §3.3 shows pre-fusion costs ~3× FLOPs by skipping the `Dk`/`R` bottlenecks. For the sizes DeepSeek-V3 uses this is decisive, so we keep the weights factored like sglang-GPU.
```
Comment (JamesBrianD) [inline docs/design/MLA.md]: But that's not what tpu-inference does. Looking at flash_attn_mla.py:146-154:
```
  q_nope, q_pe = jnp.split(q, [qk_nope_head_dim], axis=2)   # already passed through q_b_proj
  q_nope = einsum("bnp,npl->bnl", q_nope, W_UK_T)           # second step: Dk → R
```

**Comment by JamesBrianD** [comment]
The core logic is clear and sound; the other maintainers and I will review it as soon as possible.
There are a few remaining tasks to address:
- Fix linting errors.
- Complete the PR title (it should specify "absorbed mode MLA") and description.
- The documentation needs to explain how the weights `w_uk`, `w_uv`, and `w_uq` are derived, as their origin is currently somewhat difficult to understand.

File: docs/design/MLA.md (hunk 2 above)
Comment (Bob-Chen222) [reply docs/design/MLA.md]: Thanks for pointing that out. Fixed in [e5a9259](https://github.com/sgl-project/sglang-jax/pull/926/commits/e5a92595624b40f702307a2bd51acc9b3e2f2368)[e5a9259](https://github.com/sgl-project/sglang-jax/pull/926/commits/e5a92595624b40f702307a2bd51acc9b3e2f2368)

File: docs/design/MLA.md (hunk 4 above)
Comment (Bob-Chen222) [reply docs/design/MLA.md]: Thanks for pointing this out. [63ed3d1](https://github.com/sgl-project/sglang-jax/pull/926/commits/63ed3d1e641138d7197afd6b6bf6a1ef16fd4a2e) fixed and also add non goals

**Comment by Bob-Chen222** [comment]
> The core logic is clear and sound; the other maintainers and I will review it as soon as possible. There are a few remaining tasks to address:
> 
> * Fix linting errors.
> * Complete the PR title (it should specify "absorbed mode MLA") and description.
> * The documentation needs to explain how the weights `w_uk`, `w_uv`, and `w_uq` are derived, as their origin is currently somewhat difficult to understand.

All fixed. I put the notation meaning in the appendix. Thanks for the feedback

File: python/sgl_jax/srt/layers/attention/mla.py
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/mla.py]: I think that, for the initial version, we should retain the existing MLA implementation—which expands using a native approach—and then implement a separate "absorbed" MLA class.

Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/mla.py]: When invoking sglang-style attention, the declared type is always Radix Attention. The specific attention backend is then selected during the forward pass of Radix Attention. Like:
```
# python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py
# Declaration
        self.attn_mqa = RadixAttention(
            self.num_local_heads,
            self.kv_lora_rank + self.qk_rope_head_dim,
            self.scaling,
            num_kv_heads=1,
            layer_id=layer_id,
            v_head_dim=self.kv_lora_rank,
            quant_config=quant_config,
            prefix=add_prefix("attn_mqa", prefix),
        )
# Call
        attn_output = self.attn_mqa(
            q_nope_fused,
            None,
            None,
            forward_batch,
            q_rope=q_pe_fused,
            k_rope=k_pe_fused,
            save_kv_cache=save_kv_cache,
            **(
                dict(topk_indices=topk_indices)
                if topk_indices is not None
                else {}
            ),
        )
```

File: python/sgl_jax/srt/layers/attention/mla_backend.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+                ]
+            )
+        else:
+            raise ValueError(f"Invalid forward mode: {batch.forward_mode}")
+
+        seq_lens = np.copy(batch.seq_lens)
+        num_seqs = np.sum(batch.seq_lens > 0, dtype=np.int32).reshape((1,))
+
+        if batch.forward_mode == ForwardMode.DECODE:
+            distribution = np.array([0, 0, num_seqs.item()], dtype=np.int32)
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/mla_backend.py]: Please check that the distribution during decoding should be all `num_seqs`. For example: https://github.com/sgl-project/sglang-jax/pull/934

File: python/sgl_jax/srt/mem_cache/memory_pool.py [hunk 6]
```diff
@@ -872,127 +881,166 @@ def __init__(
...
+        from sgl_jax.srt.kernels.ragged_paged_attention.util import align_to
+
         self.kv_lora_rank = kv_lora_rank
+        # Kernel requires r_dim % 128 == 0; pool stores the padded width so
+        # reads/writes stay aligned. q_pe and new_k_pe must be padded before
+        # being handed to the pool / backend.
         self.qk_rope_head_dim = qk_rope_head_dim
-        self.kv_partition_axis = kv_partition_axis
+        self.qk_rope_head_dim_padded = align_to(qk_rope_head_dim, 128)
+        self.kv_dim = kv_lora_rank + self.qk_rope_head_dim_padded
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/mem_cache/memory_pool.py]: Is it possible that `kv_lora_rank` is not a multiple of 128?

File: python/sgl_jax/srt/mem_cache/memory_pool.py [hunk 7]
```diff
@@ -872,127 +881,166 @@ def __init__(
...
+        buffer_shape = (self.size + self.page_size, 1, self.kv_dim)
+        total_memory_per_layer = (
+            buffer_shape[0] * buffer_shape[1] * buffer_shape[2] * jnp.dtype(self.dtype).itemsize
+        )
+        logger.info(
+            "Total MLA KV cache memory per layer: %.4f GB, dtype: %s",
+            total_memory_per_layer / GB,
+            self.dtype,
+        )
+
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/mem_cache/memory_pool.py]: Using a 4D KV cache should be more friendly to MLA kernels.

File: docs/design/MLA.md [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+
+## 4. Work breakdown
+
+- [ ] `MLATokenToKVPool` in [memory_pool.py](../../python/sgl_jax/srt/mem_cache/memory_pool.py).
+- [ ] `MLAAttentionBackend` + `MLAAttentionMetadata` in `sglang-jax/python/sgl_jax/srt/layers/attention/mla_backend.py`.
+- [ ] Expose `W_UQ_nope`, `W_UQ_rope`, `W_UK`, `W_UV`, `W_O` as separate per-head params on `MLAAttention` (no fusion at load time).
+- [ ] Rewrite `MLAAttention.__call__` to the two-step factored absorb path (§3.9): `W_UQ_nope → W_UK` on Q, `W_UV → W_O` on output.
+- [ ] Wire dispatch in `model_runner.py` (§3.10).
+- [ ] Pad `r_dim=64` → 128 inside `MLAAttentionBackend`.
+- [ ] Unit tests for pool (shape, set/get, pytree round-trip, `get_kv_size_bytes`) and backend (kernel call, shard_map specs, decode + extend parity vs. a naive reference).
```
Comment (JamesBrianD) [inline docs/design/MLA.md:328]: A unit test needs to be added—similar to `test_flashattention.py`—to verify that the MLA backend is compatible with forwardbatch passes and produces accurate results.

Comment (Bob-Chen222) [reply docs/design/MLA.md:328]: done :)

**Review by JamesBrianD — APPROVED** [review]
LGTM
